### PR TITLE
266 constant assignment operator for customindexarray

### LINKED ISSUE #266: CustomIndexArray should allow to set the value for all compartments without a loop
This should work:
``` 
CustomIndexArray<double, AgeGroup> array;
array = 0.5 //set same value for all age groups
```
instead of 
```
for (auto age = AgeGroup(0); age < ...; ++age) {
    array[age] = 0.5;
}
```

Maybe also allow setting the value for a slice:
```
CustomIndexArray<double, AgeGroup, Gender> array;
array[AgeGroup(5)] = 0.5; //set same value for all genders
```

### PR DESCRIPTION
**Pull request by dabele**
## Changes

Added a scalar assignment operator for `CustomIndexArray`, i.e. assign every element of the array the same value in one statement, `array = 0.5`, instead of using a loop or copy assignment `array = CustomIndexArray<....>({dimensions}, 0.5)`. Also works for slices of arrays, i.e. array.slice(i) = 0.5;

Does not currently work in python, since python doesn't allow overloading assignment operators, see #279.

Also slightly improved slicing of customindexarray for single indices in one dimension

## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Tests for new functionality has been added
* [x] A local test was succesful
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)
* [ ] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [ ] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [ ] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [ ] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [ ] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [ ] There are appropriate **unit tests** and they pass.
* [ ] The git history is clean and linearized for the merge request.
* [ ] Coverage report for new code is acceptable. 

Closes #266

### CODE REVIEW

File: cpp/memilio/utils/custom_index_array.h [hunk 1]
```diff
@@ -240,6 +245,29 @@ class CustomIndexArray
...
+     * @param scalar scalar value.
+     */
+    CustomIndexArray& operator=(const Type& scalar)
+    {
+        m_y = scalar;
+        return *this;
+    }
+
+    /**
+     * equality comparison.
```
Comment (xsaschako) [inline cpp/memilio/utils/custom_index_array.h:259]: Nitpick: Equality

File: cpp/memilio/utils/custom_index_array.h [hunk 2]
```diff
@@ -563,30 +602,54 @@ class CustomIndexArray
     }
 
     /**
-     * @brief slice creates a slice into the multidimensional array
+     * @brief slice creates a slice into the multidimensional array.
```
Comment (xsaschako) [inline cpp/memilio/utils/custom_index_array.h:605]: In general: Is Creates lowercase? When using a period at the end of the sentence it looks weird not having a uppercase letter, but i guess it is style

File: cpp/tests/test_custom_index_array.cpp [hunk 3]
```diff
@@ -296,3 +329,19 @@ TEST(CustomIndexArray, chained_slice)
...
+    array.slice<Tag2>({1,2}) = 42.0; //sequence of 2 indices along one dimension
+    array.slice(Tag2(0)) = 17.0; //single index along one dimension
+
+    ASSERT_DOUBLE_EQ((array[{Tag0(0), Tag1(0), Tag2(0)}]), 17.0);
+    ASSERT_DOUBLE_EQ((array[{Tag0(0), Tag1(0), Tag2(1)}]), 42.0);
+    ASSERT_DOUBLE_EQ((array[{Tag0(0), Tag1(0), Tag2(2)}]), 42.0);
+    ASSERT_DOUBLE_EQ((array[{Tag0(0), Tag1(1), Tag2(0)}]), 17.0);
+    ASSERT_DOUBLE_EQ((array[{Tag0(0), Tag1(1), Tag2(1)}]), 42.0);
+    ASSERT_DOUBLE_EQ((array[{Tag0(0), Tag1(1), Tag2(2)}]), 42.0);
+}
```
Comment (xsaschako) [inline cpp/tests/test_custom_index_array.cpp:347]: Tests look good and "complete".

**Review by xsaschako — APPROVED** [review]
Looks good to merge!

File: cpp/memilio/utils/custom_index_array.h (hunk 2 above)
Comment (dabele) [reply cpp/memilio/utils/custom_index_array.h]: Some conventions apply of course. I don't think we have these written down anywhere and we haven't been consistent here.

I would suggest for readability, it probably should start with a capital letter. Documentation for functions should also start with verb, functions do something. Naming the function or class in the brief doc string is unnecessary.

**Review by xsaschako — APPROVED** [review]
Looks good!

**Pull request merged**